client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -336,14 +362,14 @@ public void processingFailure(Exception error) {
             state.failConnection(error);
         }
     }
-    
+
     /**
-     * @see SegmentOutputStream#write(java.nio.ByteBuffer,
-     *      java.util.concurrent.CompletableFuture)
+     * @see SegmentOutputStream#write(PendingEvent)
      */
     @Override
     @Synchronized
     public void write(PendingEvent event) throws SegmentSealedException {
```

Comment:
This class does not guarantee that they won't be. So it either needs to use synchronization to prevent it. Or specify what will happen. Because the synchronization would be complicated. I think the easiest thing to do would be make the semantic similar to flush (It will wake up the caller to prevent them from blocking) and document it. That would mean making write look like this:
```
        checkState(!state.isAlreadySealed(), "Segment: {} is already sealed", segmentName);
        ClientConnection connection;
        try {
          connection = getConnection();
        } catch (SegmentSealedException e) {
            state.addToInflight(event);
            throw e;
        }
        long eventNumber = state.addToInflight(event);
        try {
            connection.send(new Append(segmentName, writerId, eventNumber, Unpooled.wrappedBuffer(event.getData()),
                                       event.getExpectedOffset()));
        } catch (ConnectionFailedException e) {
            log.warn("Connection failed due to: ", e);
            getConnection(); // As the messages is inflight, this will perform the retransmition.
        }
```

which is not too bad and the odd behavior of adding to inflight before throwing can be explained in a comment.